lock-api/src/main/java/com/palantir/lock/v2/LeasableLockResponse.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.lock.v2;
+
+import java.util.Optional;
+
+import org.immutables.value.Value;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.annotation.JsonSerialize;
+
+@Value.Immutable
+@JsonSerialize(as = ImmutableLeasableLockResponse.class)
+@JsonDeserialize(as = ImmutableLeasableLockResponse.class)
+public abstract class LeasableLockResponse {
+    @Value.Parameter
+    public abstract Optional<LockToken> getTokenOrEmpty();
+
+    @Value.Parameter
+    public abstract Optional<Lease> getLeaseOrEmpty();
```

Comment:
`Optional<Lease>` is a bit ugly, but the other option was to add an invalid lease to timed-out responses. I picked this option.

Reply:
Two other options are to have an explicit token which contains both LockToken and Lease.

Another option is to have a union type for this; LeasableLockResponse

```
@JsonTypeInfo(type=PROPERTY, property="successful")
@JsonSubtypes(@Type(Successful.class), @Type(Unsuccessful.class)
public interface LeasableLockResponse {
    <T> T accept(Visitor<T> visitor);

    @JsonTypeName("success")
    @Value.Immutable
    @JsonDeserialize(as = ...)
    interface Successful extends LeasableLockResponse {
        LockToken getToken();
    }    

    interface Visitor<T> {
         T visit(Successful success);
         T visit(Unsuccessful failure);
    }
}
```

and then you can just visit whichever response you want; avoids having methods which might throw or whatever if it's in the wrong state?

Reply:
+1 on union type

Reply:
going with union type